gufe/chemicalstate.py

```diff
@@ -1,6 +1,76 @@
 # This code is part of OpenFE and is licensed under the MIT license.
 # For details, see https://github.com/OpenFreeEnergy/gufe
 
+from typing import Dict
+
+import numpy as np
+from frozendict import frozendict
+
+from .custom_typing import RDKitMol
+
+
 class ChemicalState:
+    """A node of an alchemical network.
+
+    Attributes
+    ----------
+    components
+        The molecular representation of the chemical state, including
+        connectivity and coordinates. This is a frozendict with user-defined
+        labels as keys, RDKit molecules as values.
+    box_vectors
+        Numpy array indicating shape and size of unit cell for the system. May
+        be a partial definition to allow for variability on certain dimensions.
+    identifier
+        Optional identifier for the chemical state; used as part of the
+        (hashable) graph node itself when the chemical state is added to an
+        `AlchemicalNetwork`
+
+    """
+
+    def __init__(
+            self,
+            components: Dict[str, RDKitMol],
```

Comment:
This is nearly what we had before but with a couple of differences. By having the key (hash) of the molecule separate there's a chance it'll get misplaced, and/or you'll end up passing around the combination of (label, Molecule) together; so we ended up lumping them together in the LigandComponent class. This has a side effect that you don't have to commit to any particular underlying representation, it's currently rdkit but once OFF is stable we could migrate if it was easier. Then it also makes it easier to attach a to_dict method to each *Component and everything is trivially serialisable. So I fought for a while to not reimplement Molecule but that's why we did in the end. Thoughts @dotsdl ?

Reply:
From today's call: we decided to proceed with the `dict` input with user-defined keys as labels for now, as these *may* be needed by an assigned edge protocol between two `ChemicalState`s to determine which components in the states map to one another (it may be possible for the protocol to discern this based on which components are different between the two states, but this may get complicated quickly for some protocols / systems). We may drop this `dict` down to a `tuple` of `Component`s later if we deem keys unnecessary here.